libs/wingsdk/src/target-sim/resource.ts

```diff
@@ -1,5 +1,308 @@
-import { ToSimulatorOutput } from "../simulator/simulator";
-import { IInflightHost, IResource } from "../std";
+import { mkdirSync, writeFileSync } from "fs";
+import { join, relative } from "path";
+import { Construct } from "constructs";
+import { SimResourceSchema } from "./schema-resources";
+import { simulatorAttrToken, simulatorHandleToken } from "./tokens";
+import { bindSimulatorResource, makeSimulatorJsClient } from "./util";
+import { App, LiftMap, Lifting } from "../core";
+import { CaseConventions, ResourceNames } from "../shared/resource-names";
+import { PolicyStatement, ToSimulatorOutput } from "../simulator";
+import {
+  IInflight,
+  IInflightHost,
+  IResource as IStdResource,
+  Resource as StdResource,
+  Node,
+  Json,
+} from "../std";
+
+/**
+ * Contract that a resource backend must implement.
+ * @inflightinterface
+ */
+export interface IResource {
+  /**
+   * Runs when the resource is started.
+   * @param context Simulator context.
+   */
+  onStart(context: IResourceContext): Promise<void>;
+
+  /**
+   * Runs when the resource is stopped.
+   */
+  onStop(): Promise<void>;
+}
+
+// TODO: make statedir a property once Wing supports properties on interfaces
+
+/**
+ * Context for implementing a simulator resource.
+ * @inflightinterface
+ */
+export interface IResourceContext {
+  /**
+   * The directory for the resource's state.
+   */
+  statedir(): string;
+
+  /**
+   * Resolves an attribute value. All attributes must be resolved during the
+   * `onStart` method.
+   *
+   * @param name The name of the attribute.
+   * @param value The value of the attribute.
+   * @inflight
+   */
+  resolveAttr(name: string, value: string): void;
+
+  /**
+   * Log a message at the current point in time. Defaults to `info` level.
+   *
+   * @param message The message to log.
+   * @param level The severity of the message.
+   * @inflight
+   */
+  log(message: string, level: LogLevel | undefined): void;
```

Comment:
Why is this needed here? Can't we just support `log()`?

Reply:
I see the argument for combining them. I'll open an a separate issue to keep tabs on this idea -- maybe we can add e.g. some kind of `levels` parameter to the built-in log function. I wanted to reduce the design surface area in this PR (there's already a lot going on) and there would might be some implementation details to work out since Node.js treats functions like `console.error` and `console.warn` the same as far as both emitting to STDERR for example.

Reply:
I agree that adding log levels to our `@log()` is out of scope for this PR (and we should add that support!).

But I am asking - if I write a resource backend and use `@log()`, where would this log go? Will it be emitted into the standard logging system?

If that's the case, I would just remove this `log()` method from `IResourceContext` for now and take care of supporting log levels through the standard logging APIs instead of introducing another API for logging just for resources.